api/src/main/java/io/opencensus/trace/BaseMessageEvent.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.trace;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Superclass for {@link MessageEvent} and {@link NetworkEvent} to resolve API backward
+ * compatibility issue.
+ *
+ * @since 0.12
+ */
+// public for io.opencensus.trace.export.SpanData.
+@SuppressWarnings("deprecation")
+public abstract class BaseMessageEvent {
+  /**
+   * Cast or convert a {@link BaseMessageEvent} to {@link MessageEvent}.
+   *
+   * <p>Warning: if the input is a {@code NetworkEvent} and contains {@code kernelTimestamp}
+   * information, this information will be dropped.
+   *
+   * @param event the {@code BaseMessageEvent} that is being cast or converted.
+   * @return a {@code MessageEvent} representation of the input.
+   * @since 0.12
+   */
+  public static MessageEvent asMessageEvent(BaseMessageEvent event) {
+    checkNotNull(event);
+    if (event instanceof MessageEvent) {
+      return (MessageEvent) event;
+    }
+    assert event instanceof NetworkEvent;
```

Comment:
I don't think the assertion is necessary, because `checkNotNull(event)` or the cast to NetworkEvent would already catch this problem without assertions enabled.

Reply:
I am trying to get rid the `BC_UNCONFIRMED_CAST` from findbugs. It cannot be suppressed by add annotation. The framework is not smart enough to tell that there are only two subclasses. Assertion is one way to work around, another one is to use judgement `if (event instanceof NetworkEvent)` to wrap the body. I'll go for the second option.

Reply:
You could suppress the warning in https://github.com/census-instrumentation/opencensus-java/blob/master/findbugs-exclude.xml.